### ISSUE TITLE
Cabal fails to solve for setup-depends due to incorrect goals

### ISSUE THREAD

**Issue by bgamari** [haskell/cabal#3617]
Cabal appears to fail to solve for `setup-depends` as it appears to be adding the package being configured to the goals list being solved for.

For instance, consider [this](https://github.com/bgamari/bytes/tree/setup-depends) branch being built with a recent GHC snapshot,
```
$ cabal build -v3
Package has never been configured. Configuring with default flags. If this
fails, please run configure manually.
Searching for ghc in path.
Found ghc at /opt/exp/ghc/roots/master/bin/ghc
("/opt/exp/ghc/roots/master/bin/ghc",["--numeric-version"])
/opt/exp/ghc/roots/master/bin/ghc is version 8.1.20160707
looking for tool ghc-pkg near compiler in /opt/exp/ghc/roots/master/bin
.
.
.
Choosing modular solver.
Resolving dependencies...
targets: bytes
constraints:
base installed (non-upgradeable package)
ghc-prim installed (non-upgradeable package)
integer-gmp installed (non-upgradeable package)
flags bytes (config file, command line flag, or user target)
stanzas bytes (config file, command line flag, or user target)
bytes ==0.15.2 (user target)
preferences:
.
.
.
strategy: PreferLatestForSelected
reorder goals: False
independent goals: False
avoid reinstalls: False
shadow packages: False
strong flags: False
max backjumps: infinite
[__0] trying: bytes-0.15.2 (user goal)
[__1] trying: base-4.9.0.0/installed-4.9... (dependency of bytes-0.15.2)
[__2] trying: bytes-setup.base~>base-4.9.0.0/installed-4.9... (dependency of
bytes-0.15.2)
[__3] trying: rts-1.0/installedrts (dependency of
base-4.9.0.0/installed-4.9...)
[__4] trying: integer-gmp-1.0.0.1/installed-1.0... (dependency of
base-4.9.0.0/installed-4.9...)
[__5] trying: ghc-prim-0.5.0.0/installed-0.5... (dependency of
base-4.9.0.0/installed-4.9...)
[__6] next goal: void (dependency of bytes-0.15.2)
[__0] fail (backjumping, conflict set: bytes, void)
Warning: solver failed to find a solution:
Could not resolve dependencies:
trying: bytes-0.15.2 (user goal)
next goal: void (dependency of bytes-0.15.2)
Dependency tree exhaustively searched.
Trying configure anyway.
Using external setup method with build-type Custom
Using explicit dependencies: False
creating dist/setup
Using Cabal library version 1.25.0.0
.
.
.
```
Compilation of `Setup.hs` then fails due to `Cabal-1.25.0.0` being used .

**Comment by ezyang** [comment]
CC @edsko @grayjay @kosmikus

**Comment by phadej** [comment]
AFAICS for old style `cabal build` solving for both `setup-depends` and package in question is valid approach (as everything have to leave in the same "universe"). Yet `cabal` is miss-behaving, it tries to build even if I add `base <0` to `setup-depends`:

```diff
% git diff
diff --git a/bytes.cabal b/bytes.cabal
index fb951a0..a8023ea 100644
--- a/bytes.cabal
+++ b/bytes.cabal
@@ -32,7 +32,7 @@ source-repository head
 
 custom-setup
   setup-depends:
-    base >= 4 && < 5,
+    base >= 4 && < 5, base <0,
```

```
ould not resolve dependencies:
trying: bytes-0.15.2 (user goal)
trying: base-4.9.0.0/installed-4.9... (dependency of bytes-0.15.2)
next goal: bytes-setup.base (dependency of bytes-0.15.2)
rejecting: bytes-setup.base~>base-4.9.0.0/installed-4.9... (conflict: bytes =>
bytes-setup.base<0)
rejecting: bytes-setup.base-4.9.0.0/installed-4.9... (multiple instances)
Dependency tree exhaustively searched.
Trying configure anyway.
```

**Comment by phadej** [comment]
`cabal new-build` seems to successfully build `Cabal-1.24` and proceed.

**Comment by grayjay** [comment]
The issue where cabal continues after failing to find a solution with the dependency solver looks like #2614.  Should it have found a solution in this case?

I got the same solver output when I tried to build `bytes` when `void` wasn't installed.  I think that is also a bug, because the warning should mention that `void` isn't available, like this:

```
Could not resolve dependencies:
trying: bytes-0.15.2 (user goal)
unknown package: void (dependency of bytes-0.15.2)
Dependency tree exhaustively searched.
```
